### PR TITLE
Linstor 4.19 fix selecting non enabled hosts

### PR DESCRIPTION
**Pull request by rp-**
### Description

This PR fixes template copy and other operations while there are nodes in the cluster under maintenance or disabled.

It was simply forgotten to check the host state is not in any none enabled state.
Additionally delete the created resource definition from Linstor if something went wrong copying.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Linstor cluster, copying a new template into primary storage and also do volume snapshots (with maintenance hosts)

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]

File: plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java [hunk 1]
```diff
@@ -958,9 +954,12 @@ private Answer copyTemplate(DataObject srcData, DataObject dstData) {
             }
             else {
                 answer = new Answer(cmd, false, "Unable to get matching Linstor endpoint.");
+                deleteResourceDefinition(pool, rscName);
             }
         } catch (ApiException exc) {
             s_logger.error("copy template failed: ", exc);
+            s_logger.info(String.format("deleting failed template resource: %s", rscName));
```
Comment (sureshanaparti) [inline plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java:961]: can move this log to `deleteResourceDefinition()` method?

Comment (rp-) [reply plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java]: done

**Review by rohityadavcloud — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/util/LinstorUtil.java [hunk 2]
```diff
@@ -107,13 +107,23 @@ public static List<String> getLinstorNodeNames(@Nonnull DevelopersApi api) throw
...
         );
-        return !sps.isEmpty() ? sps.get(0) : null;
+        return sps != null ? sps : Collections.emptyList();
+    }
+
+    public static com.linbit.linstor.api.model.StoragePool
+    getDiskfulStoragePool(@Nonnull DevelopersApi api, @Nonnull String rscName) throws ApiException
+    {
+        List<com.linbit.linstor.api.model.StoragePool> sps = getDiskfulStoragePools(api, rscName);
+        if (sps != null) {
```
Comment (sureshanaparti) [inline plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/util/LinstorUtil.java:123]:
```suggestion
        if (sps != null && !sps.isEmpty()) {
```

or use _CollectionUtils.isNotEmpty(sps)_

**Pull request merged**